xul/bootstrap.js

```diff
@@ -0,0 +1,308 @@
+
+  removeButton: function() {
+    this.tabBrowser.removeProgressListener(this);
+    this.button.parentNode.removeChild(this.button);
+    this.button = null;
+  },
+
+  onClickContext: function(aEvent) {
+    let citem = this.browserWindow.document.getElementById("esrc-explorer-item");
+    this.openExplorerOrSave(citem.getAttribute("data-url"), aEvent.ctrlKey || aEvent.metaKey);
+  },
+
+  popupShowing: function() {
+    let mrw = Services.wm.getMostRecentWindow("navigator:browser");
+    let citem = mrw.document.getElementById("esrc-explorer-item");
+    citem.hidden = true;
+    if (mrw.gContextMenu.linkURL) {
+      let srcURI = Services.io.newURI(mrw.gContextMenu.linkURL, null, null);
+      if (mrw.gBrowser.ESrcExplorer.targetUrlMatchPattern.matches(srcURI)) {
+        citem.setAttribute("data-url", mrw.gContextMenu.linkURL);
+        citem.hidden = false;
+      }
+    }
+  },
+
+  addContext: function() {
+    this.targetUrlMatchPattern = new MatchPattern([
+      '*://*/*.crx*', '*://*/*.CRX*',
+      '*://*/*.nex*', '*://*/*.NEX*',
+      '*://*/*.xpi*', '*://*/*.XPI*',
+      cws_match_pattern, ows_match_pattern,
+      amo_match_patterns, apmo_match_pattern,
+    ]);
+    let cmenu = this.browserWindow.document.getElementById("contentAreaContextMenu");
+    let citem = this.browserWindow.document.createElement("menuitem");
+    citem.setAttribute("id", "esrc-explorer-item");
+    citem.setAttribute("class", "menuitem-iconic");
+    citem.setAttribute("label", "Explore linked extension source");
+    citem.setAttribute("onclick", "gBrowser.ESrcExplorer.onClickContext(event);"); 
+    cmenu.appendChild(citem);
+    cmenu.addEventListener("popupshowing", this.popupShowing, false);
+  },
+
+  removeContext: function() {
+    let cmenu = this.browserWindow.document.getElementById("contentAreaContextMenu");
+    cmenu.removeEventListener("popupshowing", this.popupShowing);
+    let citem = this.browserWindow.document.getElementById("esrc-explorer-item");
+    cmenu.removeChild(citem);
+    this.targetUrlMatchPattern = null;
+  },
+};
+
+function BrowserWindowObserver(aHandlers) {
+  this.handlers = aHandlers;
+ }
+
+BrowserWindowObserver.prototype = {
+  observe: function(aSubject, aTopic, aData) {
+    if (aTopic == "domwindowopened") {
+      aSubject.QueryInterface(Ci.nsIDOMWindow).addEventListener("load", this, false);
+    } else if (aTopic == "domwindowclosed") {
+      if (aSubject.document.documentElement.getAttribute("windowtype") == "navigator:browser") {
+        this.handlers.onShutdown(aSubject);
+      }
+    }
+  },
+  handleEvent: function(aEvent) {
+    let aWindow = aEvent.currentTarget;
+    aWindow.removeEventListener(aEvent.type, this, false);
+
+    if (aWindow.document.documentElement.getAttribute("windowtype") == "navigator:browser") {
+      this.handlers.onStartup(aWindow);
+    }
+  }
+};
+
+function browserWindowStartup(aWindow) {
+  navigator = aWindow.navigator;
+  aWindow.gBrowser || aWindow.getBrowser();
```

Comment:
Why this? When / in which applications is `gBrowser` lazily initialized by `getBrowser`?